ubuntu 20 / llvm14

```
FAILED: tools/resgen/resgen 
: && /usr/lib/llvm-18/bin/clang++ -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -stdlib=libc++ -fPIC -fcolor-diagnostics -fvisibility=hidden -O3 -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -Wl,--gc-sections -Wl,-z,noexecstack tools/resgen/CMakeFiles/resgen.dir/src/main.cpp.o -o tools/resgen/resgen  -static-libgcc -static-libstdc++  -Wl,-Bstatic  -lc++  -lc++abi  libs/utils/libutils.a  third_party/getopt/libgetopt.a  -Wl,-Bdynamic  -pthread  -ldl  -static-libgcc -static-libstdc++  -Wl,-Bstatic  -lc++  -lc++abi  -Wl,-Bdynamic && :
/usr/bin/ld: cannot find -lc++abi
/usr/bin/ld: cannot find -lc++abi
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

Looks lib libcxx-abi isn't getting installed in llvm-14 case.  Different error with llvm-18

```
[62/1512] Linking CXX executable tools/resgen/resgen
FAILED: tools/resgen/resgen 
: && /usr/lib/llvm-18/bin/clang++ -std=c++17 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -Wno-deprecated-declarations -stdlib=libc++ -fPIC -fcolor-diagnostics -fvisibility=hidden -O3 -DNDEBUG -fomit-frame-pointer -ffunction-sections -fdata-sections -Wl,--gc-sections -Wl,-z,noexecstack tools/resgen/CMakeFiles/resgen.dir/src/main.cpp.o -o tools/resgen/resgen  -static-libgcc -static-libstdc++  -Wl,-Bstatic  -lc++  -lc++abi  libs/utils/libutils.a  third_party/getopt/libgetopt.a  -Wl,-Bdynamic  -pthread  -ldl  -static-libgcc -static-libstdc++  -Wl,-Bstatic  -lc++  -lc++abi  -Wl,-Bdynamic && :
/usr/bin/ld: cannot find -lc++abi
/usr/bin/ld: cannot find -lc++abi
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
[79/1512] Building CXX object third_party/glslang/tnt/glslang/CMakeFiles/glslang.dir/__/MachineIndependent/glslang_tab.cpp.o
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "./flutter_workspace.py", line 2802, in <module>
    main()
  File "./flutter_workspace.py", line 351, in main
    setup_platforms(platforms, github_token, cookie_file, args.plex, args.enable, args.disable, args.enable_plugin,
  File "./flutter_workspace.py", line 2197, in setup_platforms
    setup_platform(dependency, git_token, cookie_file, plex, enable, disable, enable_plugin, disable_plugin,
  File "./flutter_workspace.py", line 1997, in setup_platform
    handle_post_cmds(runtime.get('post_cmds'))
  File "./flutter_workspace.py", line 1481, in handle_post_cmds
    handle_commands_obj(post_cmd, cwd)
  File "./flutter_workspace.py", line 1556, in handle_commands_obj
    subprocess.check_call(cmd_arr, cwd=cwd, env=local_env, shell=shell_)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--config', 'Release', '--target', 'install', '--parallel']' returned non-zero exit status 1.
oem@TC4000432:~/workspace-automation$ find /usr -iname libc++*
/usr/share/doc/libc++abi1-14
/usr/share/doc/libc++-14-dev
/usr/share/doc/libc++1-14
/usr/share/doc/libc++abi-14-dev
/usr/share/lintian/overrides/libc++abi1-14
/usr/share/lintian/overrides/libc++-14-dev
/usr/share/lintian/overrides/libc++1-14
/usr/share/lintian/overrides/libc++abi-14-dev
/usr/lib/x86_64-linux-gnu/libc++abi.so.1.0
/usr/lib/x86_64-linux-gnu/libc++.so
/usr/lib/x86_64-linux-gnu/libc++.a
/usr/lib/x86_64-linux-gnu/libc++.so.1.0
/usr/lib/x86_64-linux-gnu/libc++.so.1
/usr/lib/x86_64-linux-gnu/libc++abi.so.1
/usr/lib/x86_64-linux-gnu/libc++abi.so
/usr/lib/llvm-14/lib/libc++abi.so.1.0
/usr/lib/llvm-14/lib/libc++.so
/usr/lib/llvm-14/lib/libc++abi.a
/usr/lib/llvm-14/lib/libc++.a
/usr/lib/llvm-14/lib/libc++experimental.a
/usr/lib/llvm-14/lib/libc++.so.1.0
/usr/lib/llvm-14/lib/libc++.so.1
/usr/lib/llvm-14/lib/libc++abi.so.1
/usr/lib/llvm-14/lib/libc++abi.so
```

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -43,6 +43,14 @@ jobs:
         # If llvm != 18, do --disable=toolchain-llvm-18
         run: |
           PREFER_LLVM=${{ matrix.llvm }} ./flutter_workspace.py
+          echo "Checking LLVM config..."
+          cat setup_env.sh; echo ""
+          FOUND_LLVM_VER=$(grep -oP 'export PREFER_LLVM=\K[0-9]+' setup_env.sh)
+          if [ "$FOUND_LLVM_VER" != ${{ matrix.llvm }} ]; then
+            echo "Expected LLVM version ${{ matrix.llvm }} but found ${FOUND_LLVM_VER} in setup_env.sh"
+            exit 1
+          fi
+
 
       - name: Test workspace
         run: |

--- a/flutter_workspace.py
+++ b/flutter_workspace.py
@@ -2501,82 +2501,80 @@ def get_random_mac() -> str:
     return ':'.join(map(lambda x: "%02x" % x, mac))
 
 
-env_prefix = '''#!/usr/bin/env bash -l
-
-pushd . > '/dev/null'
-SCRIPT_PATH=\"${BASH_SOURCE[0]:-$0}\"
-
-while [ -h \"$SCRIPT_PATH\" ]
-do
-    cd \"$( dirname -- \"$SCRIPT_PATH\"; )\"
-    SCRIPT_PATH=\"$( readlink -f -- \"$SCRIPT_PATH\"; )\"
-done
-cd \"$( dirname -- \"$SCRIPT_PATH\"; )\" > '/dev/null'
-
-SCRIPT_PATH=\"$( pwd; )\"
-popd  > '/dev/null'
-echo SCRIPT_PATH=$SCRIPT_PATH
-
-# LLVM/Clang environment variables
-export CC=''' + os.environ.get('CC', 'clang') + '''
-export CXX=''' + os.environ.get('CXX', 'clang++') + '''
-export LLVM_CONFIG=''' + os.environ.get('LLVM_CONFIG', 'llvm-config') + '''
-export LLVM_PREFIX=''' + os.environ.get('LLVM_PREFIX', '/usr') + '''
-export LLVM_VERSION=''' + os.environ.get('LLVM_VERSION', '') + '''
-export LLVM_CMAKEDIR=''' + os.environ.get('LLVM_CMAKEDIR', '') + '''
-export LLVM_STRIP=''' + os.environ.get('LLVM_STRIP', '/usr/bin/llvm-strip') + '''
-export PREFER_LLVM=''' + os.environ.get('PREFER_LLVM', '') + '''
-# (alias clang tools to match version)
-alias clang=clang-${LLVM_VERSION}
-alias clang++=clang++-${LLVM_VERSION}
-alias clang-tidy=${LLVM_PREFIX}/bin/clang-tidy
-alias clang-format=${LLVM_PREFIX}/bin/clang-format
-alias llvm-config=${LLVM_CONFIG}
-
-export FLUTTER_WORKSPACE=$SCRIPT_PATH
-export PATH=$FLUTTER_WORKSPACE/flutter/bin:$PATH
-export PUB_CACHE=$FLUTTER_WORKSPACE/.config/flutter_workspace/pub_cache
-export XDG_CONFIG_HOME=$FLUTTER_WORKSPACE/.config/flutter
-
-echo \"********************************************\"
-echo \"* Setting FLUTTER_WORKSPACE to:\"
-echo \"* ${FLUTTER_WORKSPACE}\"
-echo \"********************************************\"
-
-flutter doctor -v
-flutter custom-devices list
-'''
-
-env_prefix_win = r'''# PowerShell version of setup_env.sh for Windows
-
-# Get the directory of this script
-$SCRIPT_PATH = Split-Path -Parent $MyInvocation.MyCommand.Definition
-
-# Remove trailing backslash if present
-if ($SCRIPT_PATH.EndsWith('\')) {
-    $SCRIPT_PATH = $SCRIPT_PATH.Substring(0, $SCRIPT_PATH.Length)
-}
-
-$env:FLUTTER_WORKSPACE = $SCRIPT_PATH
-$env:PATH = "$env:FLUTTER_WORKSPACE\\flutter\\bin;$env:PATH"
-$env:PUB_CACHE = "$env:FLUTTER_WORKSPACE\\.config\\flutter_workspace\\pub_cache"
-
-Write-Host "********************************************"
-Write-Host "* Setting FLUTTER_WORKSPACE to:"
-Write-Host "* $env:FLUTTER_WORKSPACE"
-Write-Host "********************************************"
-
-flutter doctor -v
-flutter custom-devices list
-'''
-
 def write_env_script_header(workspace):
     """ Create environmental variable bash script """
     if sys.platform.startswith('win'):
+        env_prefix_win = r'''# PowerShell version of setup_env.sh for Windows
+
+        # Get the directory of this script
+        $SCRIPT_PATH = Split-Path -Parent $MyInvocation.MyCommand.Definition
+
+        # Remove trailing backslash if present
+        if ($SCRIPT_PATH.EndsWith('\')) {
+            $SCRIPT_PATH = $SCRIPT_PATH.Substring(0, $SCRIPT_PATH.Length)
+        }
+
+        $env:FLUTTER_WORKSPACE = $SCRIPT_PATH
+        $env:PATH = "$env:FLUTTER_WORKSPACE\\flutter\\bin;$env:PATH"
+        $env:PUB_CACHE = "$env:FLUTTER_WORKSPACE\\.config\\flutter_workspace\\pub_cache"
+
+        Write-Host "********************************************"
+        Write-Host "* Setting FLUTTER_WORKSPACE to:"
+        Write-Host "* $env:FLUTTER_WORKSPACE"
+        Write-Host "********************************************"
+
+        flutter doctor -v
+        flutter custom-devices list
+        '''
         environment_script = os.path.join(workspace, 'setup_env.ps1')
         with open(environment_script, 'w+', encoding="utf-8") as script:
             script.write(env_prefix_win)
     else:
+        env_prefix = '''#!/usr/bin/env bash -l
+        pushd . > '/dev/null'
+        SCRIPT_PATH=\"${BASH_SOURCE[0]:-$0}\"
+
+        while [ -h \"$SCRIPT_PATH\" ]
+        do
+            cd \"$( dirname -- \"$SCRIPT_PATH\"; )\"
+            SCRIPT_PATH=\"$( readlink -f -- \"$SCRIPT_PATH\"; )\"
+        done
+        cd \"$( dirname -- \"$SCRIPT_PATH\"; )\" > '/dev/null'
+
+        SCRIPT_PATH=\"$( pwd; )\"
+        popd  > '/dev/null'
+        echo SCRIPT_PATH=$SCRIPT_PATH
+
+        # LLVM/Clang environment variables
+        export CC=''' + os.environ.get('CC', 'clang') + '''
+        export CXX=''' + os.environ.get('CXX', 'clang++') + '''
+        export LLVM_CONFIG=''' + os.environ.get('LLVM_CONFIG', 'llvm-config') + '''
+        export LLVM_PREFIX=''' + os.environ.get('LLVM_PREFIX', '/usr') + '''
+        export LLVM_VERSION=''' + os.environ.get('LLVM_VERSION', '') + '''
+        export LLVM_CMAKEDIR=''' + os.environ.get('LLVM_CMAKEDIR', '') + '''
+        export LLVM_STRIP=''' + os.environ.get('LLVM_STRIP', '/usr/bin/llvm-strip') + '''
+        export PREFER_LLVM=''' + os.environ.get('PREFER_LLVM', '') + '''
+        # (alias clang tools to match version)
+        alias clang=clang-${LLVM_VERSION}
+        alias clang++=clang++-${LLVM_VERSION}
+        alias clang-tidy=${LLVM_PREFIX}/bin/clang-tidy
+        alias clang-format=${LLVM_PREFIX}/bin/clang-format
+        alias llvm-config=${LLVM_CONFIG}
+
+        export FLUTTER_WORKSPACE=$SCRIPT_PATH
+        export PATH=$FLUTTER_WORKSPACE/flutter/bin:$PATH
+        export PUB_CACHE=$FLUTTER_WORKSPACE/.config/flutter_workspace/pub_cache
+        export XDG_CONFIG_HOME=$FLUTTER_WORKSPACE/.config/flutter
+
+        echo \"********************************************\"
+        echo \"* Setting FLUTTER_WORKSPACE to:\"
+        echo \"* ${FLUTTER_WORKSPACE}\"
+        echo \"********************************************\"
+        
+        flutter doctor -v
+        flutter custom-devices list
+        '''
+
         environment_script = os.path.join(workspace, 'setup_env.sh')
         with open(environment_script, 'w+', encoding="utf-8") as script:
             script.write(env_prefix)

--- a/flutter_workspace.py
+++ b/flutter_workspace.py
@@ -2025,7 +2025,8 @@ def get_llvm_version(llvm_config):
 
     try:
         result = subprocess.run([llvm_config, '--version'], capture_output=True, text=True, check=True)
-        version = result.stdout.strip()
+        version = result.stdout.strip() # this reports full version, e.g. "16.0.6"
+        version = version.split('.')[0]
         return version
     except subprocess.CalledProcessError as e:
         print(f"Error occurred: {e}")

